### PR TITLE
pool: Don't give chainstate entire process cancel.

### PR DESCRIPTION
**Pull request by davecgh**
The pool is currently designed to shutdown the entire process deep down in the chainstate when some errors it deems to be fatal occur.  It accomplishes this by passed the cancel func for the entire process all the way down to the chainstate.  Unfortunately, this is not a good design because goroutines deep in the guts of subsystems should not have the ability to directly pull the rug out from under the under process without the upper layers having a chance to do anything about it.

Moreover, some of those errors could actually be temporary errors such as temporary loss of connection to the database or dcrd which then result in the entire process shutting down.

This is a first pass at improving the situation slightly by no longer passing the entire process's cancel func down the chainstate handler and instead arranges for the chainstate handler to return an error back to the hub's Run method which itself creates its own child context that is then canceled to cause the hub to shutdown.

It also corrects a few other issues along the way such as:

- some errors did not have the correct parameters
- the notification callback handlers were not respecting the context which could lead to goroutine leaks
- receivers should never close channels, only senders once they are sure there are no more goroutines that can write to the channel

For now, the same process shutdown behavior is maintained by having the main process shut itself down upon observing the hub shutdown which ensures all other subsystems shutdown properly too.

In other words, the overall behavior is the same, aside from the aforementioned fixes, but this change makes it feasible to further improve the behavior in the future so the pool won't just die due to what are likely temporary failures.

### CODE REVIEW

File: dcrpool.go [hunk 1]
```diff
@@ -216,7 +216,20 @@ func realMain() error {
...
 	wg.Add(2)
 	go func() {
+		// Ensure the overall process context is cancelled once Run returns
+		// since the pool can't operate without it and it's possible the hub
+		// shut itself down due to an error deep in the chain state.
+		//
+		// The entire pool really generally shouldn't be shutting itself down
+		// due to chainstate issues and should instead go into a temporarily
+		// disabled state where it stops serving miners while it retries with
+		// increasing backoffs while it attempt to resolve whatever caused it to
```
Comment (jholdstock) [inline dcrpool.go:226]:
```suggestion
		// increasing backoffs while it attempts to resolve whatever caused it to
```

Comment (davecgh) [reply dcrpool.go]: Updated.

File: pool/chainstate_test.go [hunk 2]
```diff
@@ -231,7 +231,10 @@ func testChainState(t *testing.T) {
 	var wg sync.WaitGroup
 	wg.Add(1)
 	go func() {
-		cs.handleChainUpdates(ctx)
+		err := cs.handleChainUpdates(ctx)
+		if err != nil {
+			t.Logf("unexpected error from chain update handler: %v", err)
```
Comment (jholdstock) [inline pool/chainstate_test.go:236]: Why not `t.Fatalf` here?

File: pool/hub.go [hunk 3]
```diff
@@ -623,14 +631,24 @@ func (h *Hub) shutdown() {
...
 		wg.Done()
 	}()
 	go func() {
-		h.chainState.handleChainUpdates(ctx)
+		err := h.chainState.handleChainUpdates(ctx)
+		if err != nil {
+			// Ensure the context is canceled so the remaining goroutines exit
+			// when there was an error that caused the chain update handler to
+			// exit prematurely.
+			log.Error(err)
```
Comment (jholdstock) [inline pool/hub.go:649]: Maybe skip this log if `errors.Is(err, context.Canceled)`?

File: pool/chainstate_test.go (hunk 2 above)
Comment (davecgh) [reply pool/chainstate_test.go:236]: You can't invoke `t.Fatalf` from goroutines other than the main test goroutine.

File: pool/hub.go (hunk 3 above)
Comment (davecgh) [reply pool/hub.go:649]: `handleChainUpdates` doesn't return `ctx.Err()` since it's expected that the context will be closed.  Since it doesn't return that, it can never be `context.Canceled`.

Comment (jholdstock) [reply pool/hub.go:649]: That is true if `handleChainUpdates` returns via the `case <-ctx.Done():` path, but it might return via another path.

E.g. if the ctx is canceled during the call to `cs.cfg.GetBlock(ctx, &header.PrevBlock)`, that func might return `context.Canceled`, which in turn will get returned out of `handleChainUpdates`.

File: pool/chainstate_test.go (hunk 2 above)
Comment (jholdstock) [reply pool/chainstate_test.go:236]: Hm. So an error returned by `handleChainUpdates` will cause the test to time out (which results in a panic). We don't specify a timeout so the default of 10 minutes will be used. I guess that's fine because it will still get caught by `run_tests.sh` and thus CI, but its kind of annoying if you hit it during development. Perhaps an `os.Exit(1)`?  Its a bit heavy handed, but I dont think it matters given that this is test code and we already know the test is failing.

Comment (davecgh) [reply pool/chainstate_test.go:237]: I changed it over to send the error to a buffered channel that the test checks at the end and fatals if it's non-nil.  In practice, if it exits early the test will failure due to other reasons, so it isn't strictly necessary, but it's safer to check the error.

Comment (davecgh) [reply pool/chainstate_test.go:236]: I updated the tests to respect the context when sending and receiving from the channel so that it will allow fatal test errors immediately upon context cancel.  This approach is more robust against other types of failures as well.

As an aside, tests should almost never be blindly sending to and receiving from internal channels without also selecting across the context.

Comment (jholdstock) [reply pool/chainstate_test.go:236]: Nice!

**Review by jholdstock — APPROVED** [review]
